### ISSUE TITLE
Ability to customize formatting of payloads

### ISSUE THREAD

**Issue by ndreynolds** [api-hogs/bureaucrat#32]
Currently in the default and swagger writers, the request and response bodies are just assumed to be JSON and passed through the `Poison.encode/1` with the `pretty: true` option. 

In order to support other use cases (e.g., XML, other JSON libs) or just more control of the formatting, it would be great if it were possible to configure this without re-implementing the entire writer.

This could be done via options to the writer:

```elixir
config :bureaucrat,
  writer: {Bureaucrat.MarkdownWriter, payload_formatter: MyHelper}
```

Along with a default implementation based on the current strategy.  

Also open to other suggestions.  I can work on a PR if that's something you'd be willing to include.

**Comment by mbuhot** [comment]
How about using the `content-type` header of the response to select the formatter?
It could be made pluggable if necessary, mapping mime types to the formatter function.

**Comment by ndreynolds** [comment]
Yeah, I think selecting based on the `content-type` would make a lot of sense.  Would also nicely enable mixed content types within an API.  I'll experiment with it some and try to come up with a PR.

**Comment by OpakAlex** [comment]
i think it's done?

**Comment by jangnezda** [comment]
Slightly tangential - had issue when trying to document a file upload endpoint. The swagger library can be used normally for file uploads and the resulting swagger file is correct. However, various writers in `bureaucrat` fail when running tests, because they want to json encode a custom struct (`Plug.Upload`), example with Jason encoder: 

```
** (Protocol.UndefinedError) protocol Jason.Encoder not implemented for %Plug.Upload{content_type: "audio/mp4", filename: "sample.m4a", path: "test/fixtures/sample.m4a"} of type Plug.Upload (a struct), Jason.Encoder protocol must always be explicitly implemented.
```

`Plug.Upload` is nothing fancy, just a struct with a couple of fields and for anyone wondering how to fix it, add following to your test setup/helpers code:

```elixir
  defimpl Jason.Encoder, for: Plug.Upload do
    def encode(value, opts) do
      Jason.Encode.map(Map.from_struct(value), opts)
    end
  end
```

**Comment by OpakAlex** [comment]
nice tip, thanks!
